test: increase coverage of timers

DavidCai1993 opened the pull request:
<!--
Thank you for your pull request. Please provide a description above and review
the requirements below.

Bug fixes and new features should include tests and possibly benchmarks.

Contributors guide: https://github.com/nodejs/node/blob/master/CONTRIBUTING.md
-->

##### Checklist
<!-- Remove items that do not apply. For completed items, change [ ] to [x]. -->

- [x] `make -j4 test` (UNIX), or `vcbuild test` (Windows) passes
- [x] tests and/or benchmarks are included
- [x] commit message follows commit guidelines

##### Affected core subsystem(s)
<!-- Provide affected core subsystem(s) (like doc, cluster, crypto, etc). -->
test

mscdex commented:
I think we should probably restrict the ES6 usage, otherwise we can't backport the changes easily (especially to v4.x -- which has spread and other ES6 features behind a flag yet).

Fishrock123 commented on test/parallel/test-timers-args.js:
```diff
@@ -3,29 +3,42 @@ require('../common');
 
-timeout(0);
-interval(0);
+function immediate(nargs) {
```
IMO this should be in a separate file that is about Immediates rather than Timeouts.

Sure, they are from the same module, but they share pretty much 0 implementation at the current time.

DavidCai1993 commented:
@Fishrock123 Thanks for the review, I moved that test to `test-timers-immediate.js`, PTAL.
@mscdex I agree and I changed `spread` back to `apply`. May be we should mention this in https://github.com/nodejs/node/blob/master/doc/guides/writing-tests.md#esnext-features , now in the  guide it said `"when writing tests, it is encouraged to use ES.Next features that have already landed in the ECMAScript specification"`.

mscdex commented on test/parallel/test-timers-clearImmediate.js:
```diff
@@ -2,17 +2,39 @@
+  const N = 3;
+  let count = 0;
+  function next() {
+    const immediate = setImmediate(function() {
```
Perhaps we can drop `count` entirely and just wrap the callback in a `common.mustCall()`?

mscdex commented on test/parallel/test-timers-clearImmediate.js:
```diff
@@ -2,17 +2,39 @@
+  const N = 3;
+  let count = 0;
+  function next() {
+    const immediate = setImmediate(function() {
```
Now that we have a `common.mustNotCall()`, I think we can use that here and drop `count`.

mscdex commented on test/parallel/test-timers-immediate.js:
```diff
@@ -14,17 +11,19 @@ setImmediate(common.mustCall(function() {
-  assert.deepStrictEqual(immediateC, [1, 2, 3], 'immediateC args should match');
-  assert.deepStrictEqual(immediateD, [1, 2, 3, 4, 5], '5 args should match');
-});
+  process.on('exit', function() {
```
Perhaps we can remove this event handler and just move the assertion to inside `callback`?

DavidCai1993 commented:
@mscdex Thanks for the review, PTAL.

mscdex commented:
CI: https://ci.nodejs.org/job/node-test-pull-request/6350/

mscdex commented:
CI is green, LGTM.

jasnell commented:
@Fishrock PTAL

joyeecheung commented:
@DavidCai1993 

> I agree and I changed spread back to apply. May be we should mention this in https://github.com/nodejs/node/blob/master/doc/guides/writing-tests.md#esnext-features , now in the guide it says "when writing tests, it is encouraged to use ES.Next features that have already landed in the ECMAScript specification". If more people agree with it, I think I can also submit another PR to do this.

Oops, it was me who wrote this line...forgot about the backportability part(:facepalm) +1 on fixing this.

BridgeAR commented:
This needs a rebase @DavidCai1993 

@Fishrock123 I dismiss your request as that was addressed by @DavidCai1993. Please have another look as soon as he rebased.

BridgeAR commented:
Closing due to a long inactivity. @DavidCai1993 please feel free to reopen if you want to follow up on this.
